numpy/lib/format.py

```diff
@@ -915,3 +919,151 @@ def _read_bytes(fp, size, error_template="ran out of data"):
         raise ValueError(msg % (error_template, size, len(data)))
     else:
         return data
+
+def header_tuple_dict(tuple_in):
+    return {
+        'shape': tuple_in[0],
+        'fortran_order': tuple_in[1],
+        'descr': dtype_to_descr(tuple_in[2])
+    }
+
+def has_fortran_order(arr):
+    return not arr.flags.c_contiguous and arr.flags.f_contiguous
+
+def peek(fp, length):
+    pos = fp.tell()
+    tmp = fp.read(length)
+    fp.seek(pos)
+    return tmp
+
+class NpyAppendArray:
+    def __init__(self, filename):
+        self.filename = filename
+        self.fp = None
+        self.__is_init = False
+        if os.path.isfile(filename):
+            self.__init()
+
+    def __init(self):
+        self.fp = open(self.filename, mode="rb+")
+        fp = self.fp
+
+        magic = read_magic(fp)
+        self.is_version_1 = magic[0] == 1 and magic[1] == 0
+        self.is_version_2 = magic[0] == 2 and magic[1] == 0
+
+        if not self.is_version_1 and not self.is_version_2:
+            raise NotImplementedError(
+                "version (%d, %d) not implemented" % magic
+            )
+
+        self.header_length, = unpack("<H", peek(fp, 2)) if self.is_version_1 \
+            else unpack("<I", peek(fp, 4))
+
+        self.header = read_array_header_1_0(fp) if \
+            self.is_version_1 else read_array_header_2_0(fp)
+
+        if self.header[1] != False:
+            raise NotImplementedError("fortran_order not implemented")
+
+        fp.seek(0)
+
+        self.header_bytes = fp.read(self.header_length + (
+            10 if self.is_version_1 else 12
+        ))
```

Comment:
What are 10 and 12 here?

Reply:
This is the byte offset. The header length is a little-endian unsigned short int, which has two bytes in version 1 and 4 bytes in version 2. This is where the 10 and 12 are from. I can replace this with a calculation like in
https://numpy.org/devdocs/reference/generated/numpy.lib.format.html#format-version-1-0

Reply:
Should disappear in the new version.